src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java

```diff
@@ -209,24 +209,27 @@ void invokeRequestHandler(RequestCommandAndStream requestAndStream) {
 			return this.listener.sendControlCommandAndStreamAsync(listenerCommand, responseBodyBuffer, timeout)
 					.thenRun(() -> RelayLogger.logEvent("httpSendResponseFinished", this, "control", String.valueOf(responseCommand.getStatusCode())));
 		} else {
+			TimeoutHelper timeRemaining = new TimeoutHelper(timeout);
 			RelayLogger.logEvent("httpSendResponse", this, "rendezvous", String.valueOf(responseCommand.getStatusCode()));
-			this.ensureRendezvousAsync(timeout).join();
 
 			ListenerCommand listenerCommand = new ListenerCommand(null);
 			listenerCommand.setResponse(responseCommand);
 			String command = listenerCommand.getResponse().toJsonString();
 
 			// We need to respond over the rendezvous connection
-			CompletableFuture<Void> sendCommandTask = this.rendezvousWebSocket.writeAsync(command, timeout, WriteMode.TEXT)
-					.thenAccept(bytesWritten -> {
-						RelayLogger.logEvent("httpSendResponseFinished", this, "rendezvous", String.valueOf(responseCommand.getStatusCode()));
-					});
+			CompletableFuture<Void> sendCommandTask = this.ensureRendezvousAsync(timeRemaining.remainingTime())
```

Comment:
>ensureRendezvousAsync [](start = 50, length = 21)

I wonder if the fact that this else block is only hit when `this.rendezvousWebSocket != null` means ensureRendezvousAsync can be skipped?

Reply:
It appears my statement is not true. Since ensureRendezvousAsync assigns to this.rendezvousWebSocket before it is connected, and not after what I had hoped to be the case won't work.

---
In reply to: [291718830](https://github.com/Azure/azure-relay-java/pull/45#discussion_r291718830) [](ancestors = 291718830)